### PR TITLE
added jv-hw-stream solution

### PR DESCRIPTION
**Pull request by Basilius-Third**

### CODE REVIEW

File: src/main/java/practice/CandidateValidator.java [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+
+    @Override
+    public boolean test(Candidate candidate) {
+        String periodsInUkr = candidate.getPeriodsInUkr();
+        int periodsFrom = Integer.parseInt(
+                periodsInUkr.substring(0, periodsInUkr.indexOf(PERIOD_DIVIDER)));
+        int periodsTo = Integer.parseInt(
+                periodsInUkr.substring(periodsInUkr.indexOf(PERIOD_DIVIDER) + 1));
+        return candidate.getAge() >= ELIGIBLE_PRESIDENT_AGE && candidate.isAllowedToVote()
+                && candidate.getNationality().equals("Ukrainian")
```
Comment (EugeneSinica) [inline src/main/java/practice/CandidateValidator.java:19]: it's better to make "Ukrainian" a constant

File: src/main/java/practice/StreamPractice.java [hunk 2]
```diff
@@ -23,7 +33,12 @@ public int findMinEvenNumber(List<String> numbers) {
      * But before that subtract 1 from each element on an odd position (having the odd index).
      */
     public Double getOddNumsAverage(List<Integer> numbers) {
-        return 0D;
+        return IntStream
+                .range(0, numbers.size())
+                .map(i -> i % 2 == 1 ? numbers.get(i) - 1 : numbers.get(i))
+                .filter(e -> e % 2 == 1)
+                .average()
+                .orElseThrow(NoSuchElementException::new);
```
Comment (EugeneSinica) [inline src/main/java/practice/StreamPractice.java:41]:
```suggestion
                .getAsDouble();
```
cause getAsDouble() throws this exception inside itself

File: src/main/java/practice/StreamPractice.java [hunk 3]
```diff
@@ -50,7 +70,15 @@ public Double getOddNumsAverage(List<Integer> numbers) {
...
                                           int maleToAge, List<Person> peopleList) {
-        return Collections.emptyList();
+        return peopleList
+                .stream()
+                .filter(p -> p.getSex() == Person.Sex.MAN
+                        && p.getAge() >= fromAge
+                        && p.getAge() <= maleToAge
+                        || p.getSex() == Person.Sex.WOMAN
+                        && p.getAge() >= fromAge
+                        && p.getAge() <= femaleToAge)
```
Comment (EugeneSinica) [inline src/main/java/practice/StreamPractice.java:80]: probably it'd be better to take it into a separate Predicate<Person> object

File: src/main/java/practice/StreamPractice.java [hunk 4]
```diff
@@ -75,6 +108,12 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      * parametrized with Candidate in CandidateValidator.
      */
     public List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
+        CandidateValidator validate = new CandidateValidator();
```
Comment (EugeneSinica) [inline src/main/java/practice/StreamPractice.java:111]:
```suggestion
        Predicate<Candidate> validator = new CandidateValidator();
```

**Review by EugeneSinica — CHANGES_REQUESTED** [review]
Well done, Mr. Hierarchy

**Review by EugeneSinica — APPROVED** [review]
♥

File: src/main/java/practice/StreamPractice.java [hunk 5]
```diff
@@ -14,7 +18,13 @@
...
      */
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers
+                .stream()
+                .flatMap(n -> Arrays.stream(n.split(",")))
+                .mapToInt(Integer::parseInt)
+                .filter(n -> n % 2 == 0)
+                .min()
+                .orElseThrow(() -> new RuntimeException("Can't get min value from list"));
```
Comment (Sofasmile) [inline src/main/java/practice/StreamPractice.java:27]:
```suggestion
                .orElseThrow(() -> new RuntimeException("Can't get min value from list " + numbers));
```

File: src/main/java/practice/StreamPractice.java [hunk 6]
```diff
@@ -50,7 +70,16 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      */
     public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
                                           int maleToAge, List<Person> peopleList) {
-        return Collections.emptyList();
+        Predicate<Person> filterWorkablePerson = person -> person.getSex() == Person.Sex.MAN
+                && person.getAge() >= fromAge
```
Comment (Sofasmile) [inline src/main/java/practice/StreamPractice.java:74]: no need to repeat fromAge comparison twice, it's the same for both genders

**Review by Sofasmile — CHANGES_REQUESTED** [review]
great!

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**